cpp/include/cuvs/neighbors/vamana.hpp

```diff
@@ -0,0 +1,270 @@
+ */
+
+#pragma once
+
+#include "common.hpp"
+#include <cuvs/distance/distance.hpp>
+#include <cuvs/neighbors/common.hpp>
+#include <raft/core/device_mdspan.hpp>
+#include <raft/core/host_device_accessor.hpp>
+#include <raft/core/host_mdspan.hpp>
+#include <raft/core/mdspan.hpp>
+#include <raft/core/resource/stream_view.hpp>
+#include <raft/core/resources.hpp>
+#include <raft/util/integer_utils.hpp>
+#include <rmm/cuda_stream_view.hpp>
+
+#include <optional>
+#include <variant>
+
+namespace cuvs::neighbors::vamana {
+/**
+ * @defgroup vamana_cpp_index_params Vamana index build parameters
+ * @{
+ */
+
+/**
+ * @brief ANN parameters used by VAMANA to build index
+ *
+ */
+struct index_params : cuvs::neighbors::index_params {
+  /** Degree of output graph. */
+  uint32_t graph_degree = 32;
+  /** Maximum number of visited nodes per search **/
+  uint32_t visited_size = 64;
+  /** Number of Vamana iterations. */
+  uint32_t vamana_iters = 1;
+  /** Alpha for pruning parameter */
+  float alpha = 1.2;
+  /** Maximum fraction of dataset inserted per batch.              *
+   * Larger max batch decreases graph quality, but improves speed */
+  float max_fraction = 0.06;
+  /** Base of growth rate of batch sies **/
+  float batch_base = 2;
+  /** Size of candidate queue structure - should be (2^x)-1 */
+  uint32_t queue_size = 127;
+};
+
+/**
+ * @}
+ */
+
+static_assert(std::is_aggregate_v<index_params>);
+
+/**
+ * @defgroup vamana_cpp_index Vamana index type
+ * @{
+ */
+
+/**
+ * @brief Vamana index.
+ *
+ * The index stores the dataset and the Vamana graph in device memory.
+ *
+ * @tparam T data element type
+ * @tparam IdxT type of the vector indices (represent dataset.extent(0))
+ *
+ */
+template <typename T, typename IdxT>
+struct index : cuvs::neighbors::index {
+  static_assert(!raft::is_narrowing_v<uint32_t, IdxT>,
+                "IdxT must be able to represent all values of uint32_t");
+
+ public:
+  /** Distance metric used for clustering. */
+  [[nodiscard]] constexpr inline auto metric() const noexcept -> cuvs::distance::DistanceType
+  {
+    return metric_;
+  }
+
+  /** Total length of the index (number of vectors). */
+  [[nodiscard]] constexpr inline auto size() const noexcept -> IdxT
+  {
+    auto data_rows = dataset_->n_rows();
+    return data_rows > 0 ? data_rows : graph_view_.extent(0);
+  }
+
+  /** Dimensionality of the data. */
+  [[nodiscard]] constexpr inline auto dim() const noexcept -> uint32_t { return dataset_->dim(); }
+  /** Graph degree */
+  [[nodiscard]] constexpr inline auto graph_degree() const noexcept -> uint32_t
+  {
+    return graph_view_.extent(1);
+  }
+
+  /** Dataset [size, dim] */
+  [[nodiscard]] inline auto data() const noexcept -> const cuvs::neighbors::dataset<int64_t>&
+  {
+    return *dataset_;
+  }
+
+  /** vamana graph [size, graph-degree] */
+  [[nodiscard]] inline auto graph() const noexcept
+    -> raft::device_matrix_view<const IdxT, int64_t, raft::row_major>
```

Comment:
Right, Vamana is technically not fixed degree but defines `R` as the max degree. So, I pad it to the max degree to improve memory accesses and make the algo simpler. I use `raft::upper_bound()` as a "no edge" value. In `serialize`, I condense the graph and write it out in the usual Vamana format for the DiskANN repo to use for search.

Reply:
Do you know if the average degree is significantly larger than the median? I'm just curious what the general variance looks like in a typical graph- is it a fairly fat tail? For example- can we expect a power law or small-world distribution where we have 1 or 2 vertices == R where everything else is <<R? In other words, is it okay to assume that padding is not costing a significantly high memory foot print when the data turns out to be really sparse? Mostly I'm just curious- because we do have sparse structures that could help here if the sparsity is really really high.

Reply:
I did observe average degrees to be a lot lower than the max degree for the dataset that I benchmarked DiskANN with (wiki-all-10M). In many cases, less than R/2. And it can indeed be the fat tail case. DiskANN thus uses a vector of vectors but that storage technique has its own disadvantages in comparison to the dense matrix.

Reply:
I looked at this when I was testing the serializer and the average degree is typically pretty close to R. I assume that low average degree would only be the case for very low dimension datasets, since the large hyperspace of a high-D dataset makes it very unlikely that all other nodes looked at would all violate the robustprune rule. In fact, one of the things that the gtest looks at is that the average degree > 0.75 R.

Reply:
What parameters did you see the low average degree for that dataset? I have not seen this to be the case for any of the datasets I've tested (SIFT, GIST, BIGANN, Deep 10M, random synthetic)

Reply:
I guess one thing I'm interested in knowing is how many actual zeros occur in the padded form vs the non-padded. If we're talking 70-90% zeroes then that seems like a bit of a waste to have to pad, though I do understand the perf implications here. If the sparsity is low already then padding doesn't have that much of a drawback, but if it's extremely high then I wonder if maybe we could leverage some load balancing techniques to help improve perf while still reducing the memory footprint. Anyways, things we can investigate in a future iteraiton but just wanted to learn more about the expected distribution here.

Reply:
Early in the graph construction, it's a lot of zeroes (starts with entire graph zeroed out), then adds them in batches. The final graph has not as many zeroes (in my tests < 25%, as verified by the unit tests). I've thought a bit about using a sparse structure, but the issue is that we're pretty frequently going back and reading/writing the edges of each node, as well as merging with other edge lists and sorting. A fixed degree makes the calculations easier, memory alignments nice, and fixed size sorts with CUB possible.
All this said, I don't know all the details of the raft sparse structures, so definitely something to look into in the future.

Reply:
Ah yes you are right Ben. I looked at the results for in-memory indexes again. The average degree is around 0.75 times the max degree for max degree = 64 and 32. In that case, the dense matrix is not too bad.